### ISSUE TITLE
Defining and Lowering a custom operator in ONNX-MLIR

### ISSUE THREAD

**Issue by uditagarwal97** [onnx/onnx-mlir#1496]
Hi all,

I'm working on a project that requires me to develop an alternate implementation of the convolution operator optimized for our proprietary hardware. To do this, I'm thinking of defining my custom ONNX operator 'MyConv' and then updating the ONNX IR to replace the default convolution node with MyConv. Following this, I'll have to somehow use onnx-mlir to lower the modified ONNX graph to LLVM IR.

Q1. _I'm not sure how to define my own custom ONNX operator and register it with onnx-mlir?_ onnx-mlir doc says that there is something called "mlir::ONNXCustomOp" which "allows call-out to a user-defined operation," but I couldn't find any example to use it. 

Q2. The [CustomFnTest](https://github.com/onnx/onnx-mlir/blob/main/test/onnx2mlir/CustomFnTest.cpp) in the onnx-mlir repo seems quite relevant, but is it possible to define the body of the custom operator in C++ instead?

Thanks.

**Comment by caoimhinuibrian** [comment]
@uditagarwal97 I think what you need to do is to add a custom lowering for the CONV operator to your hardware, rather than add a custm operator. The custom operator stuffis meant for adding operators that are not already supported by ONNX and convolution is already an ONNX operator. If your proprietary hardware is a new accelerator, you might take a look at the support for the NNPA accelerator (which introduces two customized dialects) and see if that supports your needs.

**Comment by uditagarwal97** [comment]
@caoimhinuibrian Thanks for your prompt reply. Sure, I'll take a look at the code for NNPA accelerator.
Closing this issue till then.

**Comment by uditagarwal97** [comment]
Hi @caoimhinuibrian,

I looked at the code for NNPA accelerator; however, I still think that using custom operators would be more suitable for our use case because:

1. Our hardware is not fixed - convolutions and matrix multiplications can be done on CPU, GPU, or Systolic Arrays, depending on the user's runtime configurations. Moreover, we already provide an abstraction for our hardware's runtime in the form of a shared library. So, users can simply call the convolution operation from the shared library without worrying about the hardware details.
2. Therefore, by rewriting the onnx graph, I could replace convolution operations with our custom implementation, say MyConv, and while lowering MyConv, onnx-mlir could just emit a call to our shared library functions - that should do the work.
Currently, I'm struggling to define a new operator in onnx-mlir (https://github.com/onnx/onnx-mlir/blob/main/docs/ImportONNXDefs.md#import-onnx-definitions-and-support-operations) that, while lowering, simply emits a call to our shared library. Any pointers on this would be helpful. Thanks.

**Comment by caoimhinuibrian** [comment]
@uditagarwal97 The intent of the custom operator is to allow "front ends" to ONNX (like Python, PyTorch, TensorFlow, ...) to define and generate an operator that is not part of the ONNX standard. That "custom operator" can then be represented in the ONNX Dialect inside onnx-mlir (and I understand that ONNX Runtime also supports these). They then get lowered either to combinations of existing operators, or a call to a library function. That doesn't seem to be what you are trying to do. From your question I understand that you want to take a standard ONNX operation, such as convolution, and then lower that to a specialized version in some cases, but not all. This is exactly what happens in the code for NNPA. Operations which can be executed on that accelerator are lowered into calls to a special library (zDNN). Other operations that are not suited to the accelerator continue down into the Kernel Dialect and flow through the rest of the MLIR and the LLVM pipelines and generate CPU code. If you only have a few operations that you need to treat specially, and if they don't have extra requirements, you may be able to have a simpler solution than the full blown accelerator path, but its not clear that that will be trivial. Part of the reason for separating out accelerated paths from the CPU paths is so that the pipeline doesn't get cluttered with lots of tests for whether different optimizations should be invoked. I should add, that if someone wants to add a GPU code gen, that should be added as an accelerator too.

If you still want to understand how to add an operation into the ONNX dialect (almost all of these operations are automatically generated from the ONNX specification, not by hand) then @chentong319 is the person who could answer your questions.

**Comment by chentong319** [comment]
1. If your MyConv is the same as ONNX's Conv Op, what you need to do is just to change the  lowering from ONNX to Krnl without a new ONNX Op. To change the lowering, you can use NNPA as Kevin suggested.  For experiment only, you can use compiler option to select which  lowering rule from onnx to krnl for ConvOp.
2. If you want to define your new MyConv Op, you can copy the definition of Conv Op in 'src/Dialect/ONNX/ONNXOps.td.inc` into `AdditionalONNXOps.td` and change the names. You need to also copy/modify the member functions of Conv, such as shape inference and customized builders if they are kept in the definition of MyConv.  They may not be needed.
3. Are you going to create a pass to rewrite ConvOp to MyConvOp?
4. To lower the MyConv or Conv to a function call to your runtime library, you can use krnl.call, which is designed to call a runtime library easily, especially for memref parameters. It's a new krnl Op. I will write some document about it. 
5. I did not see the support of lowering ONNXCustomOp or ONNXCallOp to krnl. If you use them, you still need to work on the lowering code.

**Comment by tungld** [comment]
FYI, this is a document about how to support a new custom accelerator: https://github.com/onnx/onnx-mlir/blob/main/docs/AddCustomAccelerators.md

Looking at the explanation by @uditagarwal97, I think our accelerator framework fits well. For example, creating a separate dialect to hold operations for your accelerator, say MyDialect, and registering it to `onnx-to-krnl`, then what you need to do is:
- A graph rewriting rule to lower ONNXDialect to MyDialect, e.g. ONNXConvOp to MyConvOp
- A lowering pattern to lower MyConvOp to KrnlCallOp (this pattern will be registered in `onnx-to-krnl` as well.

The big advantage  of using the accelerator framework is that we don't need to write our own driver but using the existing driver instead.

**Comment by uditagarwal97** [comment]
Thanks @caoimhinuibrian @tungld @chentong319 for your input. Okay, I'll try to give another attempt to use onnx-mlir's accelerator framework for my use case.
Closing the issue till then.
